src/databricks/labs/ucx/progress/history.py

```diff
@@ -0,0 +1,216 @@
+        """Return the dataclass-defined fields that the record type declares, and their associated types.
+
+        If the record has a "failures" attribute this is treated specially: it is removed but we signal that it was
+        present.
+
+        Arguments:
+            klass: The record type.
+        Returns:
+            A tuple containing:
+                - A dictionary of fields to include in the object data, and their type.
+                - The type of the failures field, if present.
+        """
+        field_types = {field.name: field.type for field in dataclasses.fields(klass)}
+        failures_type = field_types.pop("failures", None)
+        if failures_type not in (None, str, list[str]):
+            msg = f"Historical record {klass} has invalid 'failures' attribute of type: {failures_type}"
+            raise TypeError(msg)
+        return field_types, failures_type
+
+    def _get_id_attribute_names(self, klazz: type[Record]) -> Sequence[str]:
+        id_attribute_names = tuple(klazz.__id_attributes__)
+        all_fields = self._field_types
+        for name in id_attribute_names:
+            id_attribute_type = all_fields.get(name, None) or self._detect_property_type(klazz, name)
+            if id_attribute_type is None:
+                raise AttributeError(name=name, obj=klazz)
+            if id_attribute_type != str:
+                msg = f"Historical record {klazz} has a non-string id attribute: {name} (type={id_attribute_type})"
+                raise TypeError(msg)
+        return id_attribute_names
+
+    def _detect_property_type(self, klazz: type[Record], name: str) -> str | None:
+        maybe_property = getattr(klazz, name, None)
+        if maybe_property is None:
+            return None
+        if not isinstance(maybe_property, property):
+            msg = f"Historical record {klazz} declares an id attribute that is not a field or property: {name} (type={maybe_property})"
+            raise TypeError(msg)
+        property_getter = maybe_property.fget
+        if not property_getter:
+            msg = f"Historical record {klazz} has a non-readable property as an id attribute: {name}"
+            raise TypeError(msg)
+        type_hints = get_type_hints(property_getter) or {}
+        try:
+            return type_hints["return"]
+        except KeyError as e:
+            msg = f"Historical record {klazz} has a property with no type as an id attribute: {name}"
+            raise TypeError(msg) from e
+
+    @classmethod
+    def _get_object_type(cls, klass: type[Record]) -> str:
+        return klass.__name__
+
+    @classmethod
+    def _as_dict(cls, record: Record) -> dict[str, Any]:
+        return dataclasses.asdict(record)
+
+    def _object_id(self, record: Record) -> list[str]:
+        return [getattr(record, field) for field in self._id_attribute_names]
+
+    @classmethod
+    def _encode_non_serializable(cls, name: str, value: Any) -> Any:
```

Comment:
Wouldn't the `name` be coverred by `json.dumps`?

Reply:
This is the top-level name, and is used as the key in the `object_data` map. These are strings, so they don't need to be JSON-encoded.

It's passed through to here to: a) allow for use in the exception message; b) if we ever need to subclass/override, this allows field-specific encoding.

Nested names (within `value`) are indeed part of the JSON-encoding.